### PR TITLE
[Merged by Bors] - feat: add econstructor tactic.

### PR DESCRIPTION
**Pull request by newell**
Add `econstructor` tactic.

### CODE REVIEW

**Comment by semorrison** [comment]
Can you remove the corresponding `syntax` line from `Mathlib/Mathport/Syntax.lean`?

**Comment by newell** [comment]
> 



> Can you remove the corresponding `syntax` line from `Mathlib/Mathport/Syntax.lean`?

I was going to do that actually but then I noticed that there were other tactics such as `ApplyWith` that still have their `syntax` lines in `Mathlib/Mathport/Syntax.lean`.  I can do a follow up cleanup PR to clean the other tactics so others aren't confused about this as I was.

**Comment by digama0** [comment]
I don't see the `applyWith` tactic in `Mathport.Syntax`. There is `applyWith'`, but this is a different thing. Are there any other tactics that you saw that are duplicated?

**Comment by newell** [comment]
Actually, I think I missed the single quote on the `ApplyWith'` so disregard the previous comment.

**Comment by newell** [comment]
@digama0 ahh we are replying at the same time, as I replied before I saw your comment.  Yes, it was my mistake.

File: Mathlib/Mathport/Syntax.lean [hunk 1]
```diff
@@ -187,7 +187,6 @@ syntax caseArg := binderIdent,+ (" :" (ppSpace (ident <|> "_"))+)?
 /- M -/ syntax (name := casesType!) "cases_type!" "*"? ppSpace ident* : tactic
 /- N -/ syntax (name := abstract) "abstract" (ppSpace ident)? ppSpace tacticSeq : tactic
 
-/- E -/ syntax (name := eConstructor) "econstructor" : tactic
```
Comment (digama0) [inline Mathlib/Mathport/Syntax.lean:190]: You also need to import the new syntax into this file

**Comment by newell** [comment]
> I don't see the `applyWith` tactic in `Mathport.Syntax`. There is `applyWith'`, but this is a different thing. Are there any other tactics that you saw that are duplicated?

Not that I was aware of...I was just trying to say that if I was going to do a cleanup branch I would look for any others.  Cheers.

File: Mathlib/Tactic/Constructor.lean [hunk 2]
```diff
@@ -17,3 +17,13 @@ elab "fconstructor " : tactic => withMainContext do
   let mvarIds' ← (← getMainGoal).constructor {newGoals := .all}
   Term.synthesizeSyntheticMVarsNoPostponing
   replaceMainGoal mvarIds'
+
+/--
+`econstructor` is like `constructor`
+(it calls `apply` using the first matching constructor of an inductive datatype)
+except only non-dependent premises are added as new goals.
+-/
+elab "econstructor " : tactic => withMainContext do
```
Comment (digama0) [inline Mathlib/Tactic/Constructor.lean:26]:
```suggestion
elab "econstructor" : tactic => withMainContext do
```

File: Mathlib/Mathport/Syntax.lean (hunk 1 above)
Comment (newell) [reply Mathlib/Mathport/Syntax.lean:190]: Done, also fixed the trailing white space suggestion.

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**